### ISSUE TITLE
为什么训练dialog 模型的时候，self-attention没有进行上三角的attention mask呢？

### ISSUE THREAD

**Issue by ChengchengDu** [yangjianxin1/GPT2-chitchat#59]
inputs = model.forward(input_ids=input_ids)
您好，我看您代码里面model.forward（）这一块没有attention mask矩阵。查看了transformers的源码，attenion_mask没有加入，这样的话和进行了casual attention mask的操作有什么区别？还有就是您注释中写道“# 注意：GPT2模型的forward()函数，是对于给定的context，生成一个token，而不是生成一串token”， 怎么理解生成一个token呢？：
    def _attn(self, q, k, v, attention_mask=None, head_mask=None):
        w = torch.matmul(q, k)
        if self.scale:
            w = w / math.sqrt(v.size(-1))
        nd, ns = w.size(-2), w.size(-1)   
        b = self.bias[:, :, ns-nd:ns, :ns]
        w = w * b - 1e4 * (1 - b)

        if attention_mask is not None:
            # Apply the attention mask
            w = w + attention_mask

        w = nn.Softmax(dim=-1)(w)
        w = self.attn_dropout(w)

        # Mask heads if we want to
        if head_mask is not None:
            w = w * head_mask

        outputs = [torch.matmul(w, v)]
        if self.output_attentions:
            outputs.append(w)
        return outputs

**Comment by xu-song** [comment]
+1

**Comment by ssssxxxsd** [comment]
看着有同样的疑问， 这个三角阵写的还真是不太明显

**Comment by yangjianxin1** [comment]
掩膜矩阵在GPT2LMHeadModel模型内部实现了，可以看看一下transformers项目的源码

**Comment by Choitsugun** [comment]
我猜提问者是想问 model.forward(input_ids=input_ids) 为什么没有设定attention_mask这个参数， 不设定的话会导致在数据预处理时填充PAD的部分也会得到注意力分数, 谁能帮解释一下。

**Comment by xiaowuhu** [comment]
那个有掩码的分支走不到，传入的参数不对
